### PR TITLE
[WIP] Remove redundant NLSolve fields from alg cache

### PR DESCRIPTION
**Pull request by kanav99**
Follow up for JuliaDiffEq/DiffEqBase.jl#288, tests pass after DiffEqBase is tagged. Now I need to remove all the redundant fields from alg cache.

### CODE REVIEW

File: src/derivative_utils.jl [hunk 1]
```diff
@@ -70,6 +82,31 @@ either ForwardDiff or finite difference will be used depending on the
...
+  if isdefined(cache, :nlsolver)
+    calc_J!(cache.nlsolver, integrator, cache, is_compos)
+  elseif isdefined(cache, :J)
+    calc_J_in_cache!(integrator, cache, is_compos)
+  else
+    error("No J found in the cache")
+  end
+end
+
+function calc_J!(nlsolver::NLSolver, integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl]: why not just pass J and have 1 function?

File: src/derivative_utils.jl [hunk 2]
```diff
@@ -70,6 +82,31 @@ either ForwardDiff or finite difference will be used depending on the
 `jac_config` of the cache.
 """
 function calc_J!(integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
+  if isdefined(cache, :nlsolver)
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl]: 1 function, just an `if` to get the `J`?

File: src/derivative_utils.jl (hunk 1 above)
Comment (kanav99) [reply src/derivative_utils.jl]: That's because we unpack uf,du1,jac_config inside these functions, passing them in function too will make a lot of arguments.

**Comment by kanav99** [comment]
ABDF2 don't converge right now for some reason.

File: src/perform_step/bdf_perform_step.jl [hunk 3]
```diff
@@ -95,8 +95,7 @@ end
     cache.dtₙ₋₁ = dtₙ
     perform_step!(integrator, cache.eulercache, repeat_step)
     cache.fsalfirstprev .= integrator.fsalfirst
-    # why?
-    # nlsolver.tmp = tmp
+    nlsolver.tmp = tmp
```
Comment (ChrisRackauckas) [inline src/perform_step/bdf_perform_step.jl]: Should be `.=`

Comment (kanav99) [reply src/perform_step/bdf_perform_step.jl]: why? In fact why are we doing this? tmp is already alias of nlsolver.tmp

Comment (ChrisRackauckas) [reply src/perform_step/bdf_perform_step.jl]: if it's already an alias then yeah, it doesn't need to be there. but I thought this was the difference between `tmp` and `cache.tmp`.

**Comment by ChrisRackauckas** [comment]
Note that these changes need to be done in StochasticDiffEq too

**Comment by kanav99** [comment]
See the last commit, that was the culprit :stuck_out_tongue:

**Comment by kanav99** [comment]
Wow this PR actually found great bugs.

File: src/derivative_utils.jl [hunk 4]
```diff
@@ -371,6 +446,35 @@ function calc_W!(integrator, cache::OrdinaryDiffEqConstantCache, dtgamma, repeat
...
+  @unpack uf = nlsolver
+  mass_matrix = integrator.f.mass_matrix
+  isarray = typeof(uprev) <: AbstractArray
+  # calculate W
+  uf.t = t
+  is_compos = typeof(integrator.alg) <: CompositeAlgorithm
+  if (f isa ODEFunction && islinear(f.f)) || (f isa SplitFunction && islinear(f.f1.f))
+    J = f.f1.f
+    W = WOperator(mass_matrix, dtgamma, J, false; transform=W_transform)
+  elseif DiffEqBase.has_jac(f)
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl]: @devmotion SDIRK tests won't pass until this matches the mutable cache one, which should be a separate PR so that we can get this one correct first.

File: src/perform_step/bdf_perform_step.jl [hunk 5]
```diff
@@ -788,7 +793,7 @@ function perform_step!(integrator,cache::QNDFCache,repeat_step=false)
     @.. ϕ += γₖ[i]*D[i]
   end
   @.. ϕ *= γ
-  tm = fill!(cache.b, zero(eltype(u)))
+  tm = fill!(nlsolver.ztmp, zero(eltype(u)))
```
Comment (ChrisRackauckas) [inline src/perform_step/bdf_perform_step.jl]: what's this about?

File: src/perform_step/sdirk_perform_step.jl [hunk 6]
```diff
@@ -367,23 +373,20 @@ end
 
   # TODO: Add extrapolation
   @.. zᵧ = zprev
-  nlsolver.z = zᵧ
-
+  z .= zᵧ
   @.. tmp = uprev + d*zprev
   nlsolver.c = γ
-  zᵧ = nlsolve!(integrator, cache)
+  zᵧ .= nlsolve!(integrator, cache)
```
Comment (ChrisRackauckas) [inline src/perform_step/sdirk_perform_step.jl]: interesting. Was there an accidental harmless aliasing here before?

File: src/caches/bdf_caches.jl [hunk 7]
```diff
@@ -8,73 +7,54 @@ end
...
-  W = oop_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
-  nlsolver = oopnlsolve(alg,u,uprev,p,t,dt,f,W,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,γ,c)
-  @getoopnlsolvefields
-  eulercache = ImplicitEulerConstantCache(uf,nlsolver)
-
-  nlsolver.γ = 2//3
-  nlsolver.c = 1
+  γ, c = 2//3, 1
+  J, W = oop_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
+  nlsolver = oopnlsolve(alg,u,uprev,p,t,dt,f,W,J,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,γ,c)
```
Comment (ChrisRackauckas) [inline src/caches/bdf_caches.jl]: bugs me that this is W,J but I don't want to break that right now.

**Comment by ChrisRackauckas** [comment]
@YingboMa take a look if you have time, this is a big one.

File: src/perform_step/bdf_perform_step.jl (hunk 5 above)
Comment (kanav99) [reply src/perform_step/bdf_perform_step.jl]: The field `b` of cache was actually aliased to be `nlsovler.ztmp` everywhere

File: src/perform_step/sdirk_perform_step.jl (hunk 6 above)
Comment (kanav99) [reply src/perform_step/sdirk_perform_step.jl]: I think there were some pointer swapping things before which were dependent on the fact that we store two references to the same array therefore they worked. Now we have single reference, I would have to see if I can re-do those tricks

File: src/caches/bdf_caches.jl (hunk 7 above)
Comment (kanav99) [reply src/caches/bdf_caches.jl]: yeah the order should be fixed, I will fix this after the PR

File: src/derivative_utils.jl (hunk 1 above)
Comment (devmotion) [reply src/derivative_utils.jl]: Speaking of arguments, is there any reason for passing around `is_compos`? Its value should only depend on `integrator`, so it seems a bit redundant?

**Comment by kanav99** [comment]
Any reason you can see why PSOS tests fail in push but pass in PR?

**Comment by ChrisRackauckas** [comment]
That's a little random. I wouldn't worry about it. Every once in awhile `push` acts weird (maybe cache issue).

**Comment by devmotion** [comment]
Just noticed: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/integrator_interface.jl#L110-L111 should be updated as well, I guess
